From 63d6590f2ef3e322c98f5fc2514b05c3d5bc5431 Mon Sep 17 00:00:00 2001
From: Cliff Smolinsky <cliffsmo@microsoft.com>
Date: Tue, 23 Apr 2024 18:31:15 +0000
Subject: [PATCH] Don't create the NotificationUIManager during shutdown.

If the browser process shuts down without ever having opened a window,
the aura shutdown code will actually try to create the
NotificationUIManager just to call Shutdown(), and creating this during
shutdown results in a crash. Instead, this change adds an
IsShuttingDown() check before creating the NotificationUIManager.

This change also adds a null check to aura's reference to account for
this situation.

Bug: 336349854
Change-Id: I28ca74225ceb49f68a34c187120ed52784b39239
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/5472437
Reviewed-by: Greg Thompson <grt@chromium.org>
Commit-Queue: Cliff Smolinsky <cliffsmo@microsoft.com>
Cr-Commit-Position: refs/heads/main@{#1291446}
---

--- a/chrome/browser/browser_process_impl.cc
+++ b/chrome/browser/browser_process_impl.cc
@@ -845,8 +845,9 @@
 NotificationUIManager* BrowserProcessImpl::notification_ui_manager() {
   DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
 #if BUILDFLAG(ENABLE_CHROME_NOTIFICATIONS)
-  if (!created_notification_ui_manager_)
+  if (!created_notification_ui_manager_ && !shutting_down_) {
     CreateNotificationUIManager();
+  }
   return notification_ui_manager_.get();
 #else
   return nullptr;
--- a/chrome/browser/lifetime/application_lifetime_aura.cc
+++ b/chrome/browser/lifetime/application_lifetime_aura.cc
@@ -44,7 +44,11 @@
   // This clears existing notifications from the message center and their
   // associated ScopedKeepAlives. Chrome OS doesn't use ScopedKeepAlives for
   // notifications.
-  g_browser_process->notification_ui_manager()->StartShutdown();
+  if (auto* notification_ui_manager =
+          g_browser_process->notification_ui_manager();
+      notification_ui_manager) {
+    notification_ui_manager->StartShutdown();
+  }
 #endif
 
   views::Widget::CloseAllSecondaryWidgets();
